### PR TITLE
[LLVM][SVE][CodeGen] Fix incorrect isel for signed saturating instructions.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
The immediate forms of SQADD and SQSUB interpret their immediate
operand as unsigned and thus effectively only support positive
immediate operands.
    
The original code is only wrong for the i8 cases because they
previously accepted all values, however, the new patterns enable more
uses and this I've added tests for the larger element types as well.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 1]
```diff
@@ -4014,6 +4020,50 @@ bool AArch64DAGToDAGISel::SelectSVEAddSubImm(SDValue N, MVT VT, SDValue &Imm,
...
+  int64_t Val = cast<ConstantSDNode>(N)
+                    ->getAPIntValue()
+                    .trunc(VT.getFixedSizeInBits())
+                    .getSExtValue();
+
+  // Signed saturating instructions treat their immediate operand as unsigned.
+  if (Val < 0)
+    return false;
+
+  switch (VT.SimpleTy) {
```
Comment (kmclaughlin-arm) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:4038]: This switch looks the same as the one in `SelectSVEAddSubImm` above, would it be possible to use the same function for the signed instructions and pass an extra `Signed` flag for checking `Val < 0`?

Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:4038]: I thought the same but whilst the switch is the same, the type of `Val` is different.  For `SelectSVEAddSubImm` it is unsigned but here it is signed.

File: llvm/test/CodeGen/AArch64/sve-intrinsics-int-arith-imm.ll [hunk 2]
```diff
@@ -1059,6 +1059,20 @@ define <vscale x 16 x i8> @sqadd_b_lowimm(<vscale x 16 x i8> %a) {
...
 }
 
+; Immediate instruction form only supports positive values.
+define <vscale x 16 x i8> @sqadd_b_negimm(<vscale x 16 x i8> %a) {
+; CHECK-LABEL: sqadd_b_negimm:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov z1.b, #-1 // =0xffffffffffffffff
+; CHECK-NEXT:    sqadd z0.b, z0.b, z1.b
+; CHECK-NEXT:    ret
+  %elt = insertelement <vscale x 16 x i8> undef, i8 -1, i32 0
```
Comment (david-arm) [inline llvm/test/CodeGen/AArch64/sve-intrinsics-int-arith-imm.ll:1069]: Perhaps I'm missing something, but the encoding of SQADD permits a 8-bit immediate that is treated as unsigned, i.e. all values in the range [0,255]. The documentation says:

`The immediate is an unsigned value in the range 0 to 255`

So I don't see why we can't lower a splat of immediates <i8 0xFF, i8 0xFF, ...> (i.e. a splat of the value 255) to the immediate form of sqadd?

**Comment by davemgreen** [comment]
This certainly sounds like one of those problems that is easy to get wrong. It sounds like the generation of the intrinsics in clang might need to change too?

**Comment by paulwalker-arm** [comment]
> This certainly sounds like one of those problems that is easy to get wrong. It sounds like the generation of the intrinsics in clang might need to change too?

Please can you elaborate?  We don't typically expose immediate forms to Clang unless that's the only instruction format available, which is not the case for sqadd/sqsub.  We leave it as a code generation optimisation to use them when appropriate.

File: llvm/test/CodeGen/AArch64/sve-intrinsics-int-arith-imm.ll (hunk 2 above)
Comment (paulwalker-arm) [reply llvm/test/CodeGen/AArch64/sve-intrinsics-int-arith-imm.ll:1069]: Thinking about it a little more we should be able to invert negative constants along with the operation (i.e. "sqadd A, #-1" -> "sqsub A, #1".  The question is whether that is something I should do within this patch? Or whether we're happy to keep the bug fix separate?

**Comment by paulwalker-arm** [comment]
Given it involved editing largely the same code I've included the changes to ensure optimal code generation. Please let me know if you prefer the bug fix to land separately.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 3]
```diff
@@ -4014,6 +4020,50 @@ bool AArch64DAGToDAGISel::SelectSVEAddSubImm(SDValue N, MVT VT, SDValue &Imm,
...
+  if (!isa<ConstantSDNode>(N))
+    return false;
+
+  SDLoc DL(N);
+  int64_t Val = cast<ConstantSDNode>(N)
+                    ->getAPIntValue()
+                    .trunc(VT.getFixedSizeInBits())
+                    .getSExtValue();
+
+  // Signed saturating instructions treat their immediate operand as unsigned.
```
Comment (david-arm) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:4039]: Is it worth expanding this comment a little to say something like

// Signed saturating instructions treat their immediate operand as unsigned,
// whereas the intrinsics are defined to work on signed operands.

Comment (paulwalker-arm) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: Done

**Review by david-arm — APPROVED** [review]
LGTM!

**Comment by davemgreen** [comment]
> > This certainly sounds like one of those problems that is easy to get wrong. It sounds like the generation of the intrinsics in clang might need to change too?
> 
> Please can you elaborate if I've misunderstood? We don't typically expose immediate forms to Clang unless that's the only instruction format available, which is not the case for sqadd/sqsub. We leave it as a code generation optimisation to use them when appropriate.

Hi. It was sqadd_n_s8 that I was thinking of, but looking again it looks like the argument is signed, not unsigned. There are some details in the documentation for the intrinsic about it producing sqadd or sqsub depending on the range of the value, so if that is now happening this sounds good to me!

**Pull request merged**